rust/src/quic/frames.rs

```diff
@@ -115,13 +122,124 @@ impl fmt::Display for StreamTag {
     }
 }
 
+#[derive(Debug, PartialEq)]
+pub(crate) struct Ack {
+    pub largest_acknowledged: u64,
+    pub ack_delay: u64,
+    pub ack_range_count: u64,
+    pub first_ack_range: u64,
+}
+
+#[derive(Debug, PartialEq)]
+pub(crate) struct Crypto {
+    pub ciphers: Vec<TlsCipherSuiteID>,
+    pub extv: Vec<QuicTlsExtension>,
+    //does not work, because of lifetime due to references to the slice used for parsing
+    //and cannot manage to create a stucture with the Vector the slices refer to...
+    //pub exts: Vec<TlsExtension>,
+}
+
 #[derive(Debug, PartialEq)]
 pub(crate) enum Frame {
     Padding,
+    Ack(Ack),
+    Crypto(Crypto),
     Stream(Stream),
     Unknown(Vec<u8>),
 }
 
+fn parse_ack_frame(input: &[u8]) -> IResult<&[u8], Frame, QuicError> {
+    let rest = input;
+    let (rest, largest_acknowledged) = quic_var_uint(rest)?;
+    let (rest, ack_delay) = quic_var_uint(rest)?;
+    let (rest, ack_range_count) = quic_var_uint(rest)?;
+    let (rest, first_ack_range) = quic_var_uint(rest)?;
+
+    if ack_range_count != 0 {
+        //TODO RFC9000 section 19.3.1.  ACK Ranges
+        return Err(nom::Err::Error(QuicError::UnhandledYet));
+    }
+
+    Ok((
+        rest,
+        Frame::Ack(Ack {
+            largest_acknowledged,
+            ack_delay,
+            ack_range_count,
+            first_ack_range,
+        }),
+    ))
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct QuicTlsExtension {
+    pub etype: TlsExtensionType,
+    pub values: Vec<u8>,
+    pub indices: Vec<usize>,
+}
+
+// get interesting stuff out of parsed tls extensions
+fn quic_get_tls_extensions(input: Option<&[u8]>) -> Vec<QuicTlsExtension> {
+    let mut extv = Vec::new();
+    if let Some(extr) = input {
+        if let Ok((_, exts)) = parse_tls_extensions(extr) {
+            for e in &exts {
```

Comment:
@chifflier do you have some idea on how to store the result of `parse_tls_extensions` into the transaction ?
The problem is that the result has references to the slice it parsed.
And I cannot manage to store the vector the slices refer to (like a self referential struct, or owning_ref stuff)

Reply:
> I think it would be much easier to store owned data in enum variants.

But I would need to redo the 28 variants.
In this case, I would better rewrite my own `parse_tls_extensions`, would I not ?

> BTW, you loose information below, for ex the SNIType is lost (not sure this is a problem, I just noticed it).

Yes that is a problem, and I only get 2 out of the 28 variants...

Reply:
@chifflier We were hoping to just store the result of `parse_tls_extensions`, which we can't do as it would become a self referencing struct, and hoping to save the effort of remapping each variant to our own owned one.

Being a library I think it makes sense what you did, had we done a similar parser internally we probably would have had the parser returned owned data as it fits our use better, I think. anyways.